File: open_spiel/python/environments/cliff_walking.py
# Copyright 2019 DeepMind Technologies Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""A cliff walking single agent reinforcement learning environment."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np

from open_spiel.python import rl_environment

# Actions
RIGHT, UP, LEFT, DOWN = range(4)


class Environment(object):
  r"""A cliff walking reinforcement learning environment.

  This is a deterministic environment that can be used to test RL algorithms.
  Note there are *no illegal moves* in this environment--if the agent is on the
  edge of the cliff and takes an action which would yield an invalid position,
  the action is ignored (as if there were walls surrounding the cliff).

  Cliff example for height=3 and width=5:

                |   |   |   |   |   |
                |   |   |   |   |   |
                | S | x | x | x | G |

  where `S` is always the starting position, `G` is always the goal and `x`
  represents the zone of high negative reward to be avoided. For this instance,
  the optimum policy is depicted as follows:

                |   |   |   |   |   |
                |-->|-->|-->|-->|\|/|
                |/|\| x | x | x | G |

  yielding a reward of -6 (minus 1 per time step).

  See pages 132 of Rich Sutton's book for details:
  http://www.incompleteideas.net/book/bookdraft2018mar21.pdf
  """

  def __init__(self, height=4, width=8, discount=1.0, max_t=100):
    if height < 2 or width < 3:
      raise ValueError("height must be >= 2 and width >= 3.")
    self._height = height
    self._width = width
    self._legal_actions = [RIGHT, UP, LEFT, DOWN]
    self._should_reset = True
    self._max_t = max_t

    # Discount returned at non-initial steps.
    self._discounts = [discount] * self.num_players

  def reset(self):
    """Resets the environment."""
    self._should_reset = False
    self._time_counter = 0
    self._state = np.array([self._height - 1, 0])

    observations = {
        "info_state": [self._state.copy()],
        "legal_actions": [self._legal_actions],
        "current_player": 0,
    }

    return rl_environment.TimeStep(
        observations=observations,
        rewards=None,
        discounts=None,
        step_type=rl_environment.StepType.FIRST)

  def step(self, actions):
    """Updates the environment according to `actions` and returns a `TimeStep`.

    Args:
      actions: A singleton list with an integer, or an integer, representing the
        action the agent took.

    Returns:
      A `rl_environment.TimeStep` namedtuple containing:
        observation: singleton list of dicts containing player observations,
            each corresponding to `observation_spec()`.
        reward: singleton list containing the reward at this timestep, or None
            if step_type is `rl_environment.StepType.FIRST`.
        discount: singleton list containing the discount in the range [0, 1], or
            None if step_type is `rl_environment.StepType.FIRST`.
        step_type: A `rl_environment.StepType` value.
    """
    if self._should_reset:
      return self.reset()
    self._time_counter += 1

    if isinstance(actions, list):
      action = actions[0]
    elif isinstance(actions, int):
      action = actions
    else:
      raise ValueError("Action not supported.", actions)

    dx = 0
    dy = 0
    if action == LEFT:
      dx -= 1
    elif action == RIGHT:
      dx += 1

    if action == UP:
      dy -= 1
    elif action == DOWN:
      dy += 1

    self._state += np.array([dy, dx])
    self._state = self._state.clip(0, [self._height - 1, self._width - 1])

    done = self._is_pit(self._state) or self._is_goal(self._state)
    done = done or self._time_counter >= self._max_t
    # Return observation
    step_type = (
        rl_environment.StepType.LAST if done else rl_environment.StepType.MID)
    self._should_reset = step_type == rl_environment.StepType.LAST

    observations = {
        "info_state": [self._state.copy()],
        "legal_actions": [self._legal_actions],
        "current_player": 0,
    }

    return rl_environment.TimeStep(
        observations=observations,
        rewards=[self._get_reward(self._state)],
        discounts=self._discounts,
        step_type=step_type)

  def _is_goal(self, pos):
    """Check if position is bottom right corner of grid."""
    return pos[0] == self._height - 1 and pos[1] == self._width - 1

  def _is_pit(self, pos):
    """Check if position is in bottom row between start and goal."""
    return (pos[1] > 0 and pos[1] < self._width - 1 and
            pos[0] == self._height - 1)

  def _get_reward(self, pos):
    if self._is_pit(pos):
      return -100.0
    else:
      return -1.0

  def observation_spec(self):
    """Defines the observation provided by the environment.

    Each dict member will contain its expected structure and shape.

    Returns:
      A specification dict describing the observation fields and shapes.
    """
    return dict(
        info_state=tuple([2]),
        legal_actions=(len(self._legal_actions),),
        current_player=(),
    )

  def action_spec(self):
    """Defines action specifications.

    Specifications include action boundaries and their data type.

    Returns:
      A specification dict containing action properties.
    """
    return dict(
        num_actions=len(self._legal_actions),
        min=min(self._legal_actions),
        max=max(self._legal_actions),
        dtype=int,
    )

  @property
  def num_players(self):
    return 1

  @property
  def is_turn_based(self):
    return False

File: open_spiel/games/cliff_walking.cc
// Copyright 2019 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "open_spiel/games/cliff_walking.h"

#include <algorithm>
#include <utility>

#include "open_spiel/game_parameters.h"
#include "open_spiel/spiel.h"
#include "open_spiel/spiel_utils.h"

namespace open_spiel {
namespace cliff_walking {
namespace {

enum CliffWalkingAction { RIGHT = 0, UP = 1, LEFT = 2, DOWN = 3 };

// Facts about the game.
const GameType kGameType{/*short_name=*/"cliff_walking",
                         /*long_name=*/"CliffWalking",
                         GameType::Dynamics::kSequential,
                         GameType::ChanceMode::kDeterministic,
                         GameType::Information::kPerfectInformation,
                         GameType::Utility::kGeneralSum,
                         GameType::RewardModel::kRewards,
                         /*max_num_players=*/1,
                         /*min_num_players=*/1,
                         /*provides_information_state_string=*/true,
                         /*provides_information_state_tensor=*/true,
                         /*provides_observation_string=*/true,
                         /*provides_observation_tensor=*/true,
                         /*parameter_specification=*/
                         {{"height", GameParameter(kDefaultHeight)},
                          {"width", GameParameter(kDefaultWidth)},
                          {"horizon", GameParameter(kDefaultHorizon)}}};

std::shared_ptr<const Game> Factory(const GameParameters& params) {
  return std::shared_ptr<Game>(new CliffWalkingGame(params));
}

REGISTER_SPIEL_GAME(kGameType, Factory);

}  // namespace

CliffWalkingState::CliffWalkingState(std::shared_ptr<const Game> game)
    : State(game) {
  const CliffWalkingGame& parent_game =
      static_cast<const CliffWalkingGame&>(*game);
  height_ = parent_game.Height();
  width_ = parent_game.Width();
  horizon_ = parent_game.MaxGameLength();
  player_row_ = parent_game.Height() - 1;
}

int CliffWalkingState::CurrentPlayer() const {
  if (IsTerminal()) return kTerminalPlayerId;
  return 0;
}

std::vector<Action> CliffWalkingState::LegalActions() const {
  if (IsTerminal()) return {};
  return {RIGHT, UP, LEFT, DOWN};
}

std::string CliffWalkingState::ActionToString(int player,
                                              Action action_id) const {
  SPIEL_CHECK_EQ(player, 0);
  switch (action_id) {
    case RIGHT:
      return "RIGHT";
    case UP:
      return "UP";
    case LEFT:
      return "LEFT";
    case DOWN:
      return "DOWN";
    default:
      SpielFatalError("Out of range action");
  }
}

std::string CliffWalkingState::ToString() const {
  std::string str;
  str.reserve(height_ * (width_ + 1));
  for (int r = 0; r < height_; ++r) {
    for (int c = 0; c < width_; ++c) {
      if (r == player_row_ && c == player_col_)
        str += 'P';
      else if (IsCliff(r, c))
        str += 'X';
      else if (IsGoal(r, c))
        str += 'G';
      else
        str += '.';
    }
    str += '\n';
  }
  return str;
}

bool CliffWalkingState::IsTerminal() const {
  return time_counter_ >= horizon_ || IsCliff(player_row_, player_col_) ||
         IsGoal(player_row_, player_col_);
}

std::vector<double> CliffWalkingState::Rewards() const {
  if (IsCliff(player_row_, player_col_)) return {-100.0};
  if (time_counter_ == 0) return {0.0};
  return {-1.0};
}

std::vector<double> CliffWalkingState::Returns() const {
  if (IsCliff(player_row_, player_col_)) return {-100.0 - time_counter_ + 1};
  return {time_counter_ * -1.0};
}

std::string CliffWalkingState::InformationStateString(Player player) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  return HistoryString();
}

std::string CliffWalkingState::ObservationString(Player player) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  return ToString();
}

void CliffWalkingState::ObservationTensor(Player player,
                                          absl::Span<float> values) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  SPIEL_CHECK_EQ(values.size(), height_ * width_);
  std::fill(values.begin(), values.end(), 0.);
  values[player_row_ * width_ + player_col_] = 1.0;
}

void CliffWalkingState::InformationStateTensor(Player player,
                                               absl::Span<float> values) const {
  SPIEL_CHECK_GE(player, 0);
  SPIEL_CHECK_LT(player, num_players_);
  SPIEL_CHECK_EQ(values.size(), kNumActions * horizon_);
  for (int i = 0; i < history_.size(); i++) {
    values[i * kNumActions + history_[i].action] = 1;
  }
}

std::unique_ptr<State> CliffWalkingState::Clone() const {
  return std::unique_ptr<State>(new CliffWalkingState(*this));
}

void CliffWalkingState::DoApplyAction(Action move) {
  switch (move) {
    case RIGHT:
      ++player_col_;
      break;
    case UP:
      --player_row_;
      break;
    case LEFT:
      --player_col_;
      break;
    case DOWN:
      ++player_row_;
      break;
    default:
      SpielFatalError("Unexpected action");
  }
  player_row_ = std::min(std::max(player_row_, 0), height_ - 1);
  player_col_ = std::min(std::max(player_col_, 0), width_ - 1);
  ++time_counter_;
}

bool CliffWalkingState::IsCliff(int row, int col) const {
  return col > 0 && col < width_ - 1 && row == height_ - 1;
}

bool CliffWalkingState::IsGoal(int row, int col) const {
  return row == height_ - 1 && col == width_ - 1;
}

CliffWalkingGame::CliffWalkingGame(const GameParameters& params)
    : Game(kGameType, params),
      height_(ParameterValue<int>("height")),
      width_(ParameterValue<int>("width")),
      horizon_(ParameterValue<int>("horizon")) {
  SPIEL_CHECK_GE(height_, 2);
  SPIEL_CHECK_GE(width_, 3);
}

}  // namespace cliff_walking
}  // namespace open_spiel
